ext/mbstring/libmbfl/filters/mbfilter_utf8.c

```diff
@@ -355,3 +580,536 @@ static zend_string* mb_cut_utf8(unsigned char *str, size_t from, size_t len, uns
+
+retry:
+	switch (filter->status & 0xff) {
+	case 0x00:
+		if (c < 0x80) {
+			CK((*filter->output_function)(c, filter->data));
+		} else if (c >= 0xc2 && c <= 0xdf) { /* 2byte code first char: 0xc2-0xdf */
+			filter->status = 0x10;
+			filter->cache = c & 0x1f;
+		} else if (c >= 0xe0 && c <= 0xef) { /* 3byte code first char: 0xe0-0xef */
+			filter->status = 0x20;
+			filter->cache = c & 0xf;
+		} else if (c >= 0xf0 && c <= 0xf4) { /* 3byte code first char: 0xf0-0xf4 */
+			filter->status = 0x30;
+			filter->cache = c & 0x7;
+		} else {
+			CK(mbfl_filt_put_invalid_char(filter));
+		}
+		break;
+
+	case 0x10: /* 2byte code 2nd char: 0x80-0xbf */
+	case 0x21: /* 3byte code 3rd char: 0x80-0xbf */
+	case 0x32: /* 4byte code 4th char: 0x80-0xbf */
+		filter->status = 0;
+		if (c >= 0x80 && c <= 0xbf) {
+			s = (filter->cache << 6) | (c & 0x3f);
+			filter->cache = 0;
+
+			if (filter->from->no_encoding == mbfl_no_encoding_utf8_docomo && mbfilter_conv_r_map_tbl(s, &s1, mbfl_docomo2uni_pua, 4)) {
+				s = mbfilter_sjis_emoji_docomo2unicode(s1, &snd);
+			} else if (filter->from->no_encoding == mbfl_no_encoding_utf8_kddi_a && mbfilter_conv_r_map_tbl(s, &s1, mbfl_kddi2uni_pua, 7)) {
+				s = mbfilter_sjis_emoji_kddi2unicode(s1, &snd);
+			} else if (filter->from->no_encoding == mbfl_no_encoding_utf8_kddi_b && mbfilter_conv_r_map_tbl(s, &s1, mbfl_kddi2uni_pua_b, 8)) {
+				s = mbfilter_sjis_emoji_kddi2unicode(s1, &snd);
+			} else if (filter->from->no_encoding == mbfl_no_encoding_utf8_sb && mbfilter_conv_r_map_tbl(s, &s1, mbfl_sb2uni_pua, 6)) {
+				s = mbfilter_sjis_emoji_sb2unicode(s1, &snd);
+			}
+
+			if (snd > 0) {
+				CK((*filter->output_function)(snd, filter->data));
+			}
+			CK((*filter->output_function)(s, filter->data));
+		} else {
+			CK(mbfl_filt_put_invalid_char(filter));
+			goto retry;
+		}
+		break;
+
+	case 0x20: /* 3byte code 2nd char: 0:0xa0-0xbf,D:0x80-9F,1-C,E-F:0x80-0x9f */
+		s = (filter->cache << 6) | (c & 0x3f);
+		c1 = filter->cache & 0xf;
+
+		if ((c >= 0x80 && c <= 0xbf) &&
+			 ((c1 == 0x0 && c >= 0xa0) ||
+			 (c1 == 0xd && c < 0xa0) ||
+			 (c1 > 0x0 && c1 != 0xd))) {
+			filter->cache = s;
+			filter->status++;
+		} else {
+			CK(mbfl_filt_put_invalid_char(filter));
+			goto retry;
+		}
+		break;
+
+	case 0x30: /* 4byte code 2nd char: 0:0x90-0xbf,1-3:0x80-0xbf,4:0x80-0x8f */
+		s = (filter->cache << 6) | (c & 0x3f);
+		c1 = filter->cache & 0x7;
+
+		if ((c >= 0x80 && c <= 0xbf) &&
+			 ((c1 == 0x0 && c >= 0x90) ||
+			 (c1 == 0x4 && c < 0x90) ||
+			 (c1 > 0x0 && c1 != 0x4))) {
+			filter->cache = s;
+			filter->status++;
+		} else {
+			CK(mbfl_filt_put_invalid_char(filter));
+			goto retry;
+		}
+		break;
+
+	case 0x31: /* 4byte code 3rd char: 0x80-0xbf */
+		if (c >= 0x80 && c <= 0xbf) {
+			filter->cache = (filter->cache << 6) | (c & 0x3f);
+			filter->status++;
+		} else {
+			CK(mbfl_filt_put_invalid_char(filter));
+			goto retry;
+		}
+		break;
+
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
```

Comment:
```suggestion
	EMPTY_SWITCH_DEFAULT_CASE();
	}
```
Which highlights an issue of the ``case`` statements not being indented.

Reply:
Hmm, both of these are in legacy code which I want to delete as soon as possible...

Reply:
ACK